File: Examples/intel_examples/examples_core_c/servicefuncsc/source/servicefuncs.c
/*******************************************************************************
* Copyright 1999-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!      mklservicefunctions example program demonstrates a lot of
!      Intel(R) MKL service functions.
!******************************************************************************/
#include <stdio.h>
#include <mkl.h>

#define NN 1000
#define buf_len 198

#if defined(MKL_ILP64) && defined(_WIN32)
#define FORMAT "%I64d"
#else
#define FORMAT "%lld"
#endif 

static void aux_print_n_threads(void);
static void aux_call_dgemm(MKL_INT N, double* A, double* B, double* C, int PRINT_CLOCKS);

int main(void) {

    double freq, seconds_s, seconds_e;

    double *a;
    double *b;
    double *c;

    int n = NN;

    MKL_INT64 allocated_bytes;
    int       allocated_buffers;

    char buf[buf_len];
    MKLVersion ver;

/** Information and defaults **/

    seconds_s = dsecnd();

    printf("\nIntel(R) MKL service functions example started\n");

    printf("\nIntel(R) MKL release version:\n");
    MKL_Get_Version_String(buf, buf_len);
    printf("%s\n",buf);

    MKL_Get_Version(&ver);
    printf("    Major version:          %d\n",ver.MajorVersion);
    printf("    Update version:         %d\n",ver.UpdateVersion);
    printf("    Product status:         %s\n",ver.ProductStatus);
    printf("    Build:                  %s\n",ver.Build);
    printf("    Platform:               %s\n",ver.Platform);
    printf("    Processor optimization: %s\n",ver.Processor);

    printf("\nInformation and defaults\n");

    freq = MKL_Get_Cpu_Frequency();
    printf("    Current CPU frequency:%8.4fGHz\n",freq);

    freq = MKL_Get_Max_Cpu_Frequency();
    printf("    Maximum CPU frequency:%8.4fGHz\n",freq);

    freq = MKL_Get_Clocks_Frequency();
    printf("    Frequency:%8.4fGHz\n",freq);

    aux_print_n_threads();

/** Memory functions **/
    MKL_Peak_Mem_Usage(MKL_PEAK_MEM_ENABLE);
    printf("\nMemory functions\n");
    printf("    Allocate DGEMM's arrays\n");

    a = (double*)MKL_malloc(n*n*sizeof(double),128);
    b = (double*)MKL_malloc(n*sizeof(double),128); /** Allocates a memory buffer of smaller size to realloc it to appropriate size later **/
    c = (double*)MKL_calloc(n*n,sizeof(double),128);

    if ( a == NULL || b == NULL || c == NULL ) {
        printf("\nAllocation of arrays failed\n");
        return 1;
    } else {
        b = (double*)MKL_realloc(b,n*n*sizeof(double));
        printf("    CALL DGEMM\n");
        aux_call_dgemm(n,a,b,c,0);
        printf("    ...Done\n");

        allocated_bytes = MKL_Peak_Mem_Usage(MKL_PEAK_MEM);
        printf("    Peak memory allocated by Intel(R) MKL allocator :"FORMAT" bytes.\n",allocated_bytes);

        allocated_bytes = MKL_Mem_Stat(&allocated_buffers);
        printf("    Currently allocated by Intel(R) MKL allocator :"FORMAT" bytes in%3d buffers.\n",allocated_bytes,(int)allocated_buffers);

        MKL_Free_Buffers();
        allocated_bytes = MKL_Mem_Stat(&allocated_buffers);
        printf("    After Mkl_Free_Buffers was called:\n");
        printf("    Currently allocated by Intel(R) MKL allocator :"FORMAT" bytes in%3d buffers.\n",allocated_bytes,(int)allocated_buffers);
        if ( MKL_Peak_Mem_Usage(MKL_PEAK_MEM_DISABLE) < 0 ) {
            printf("Peak memory statistics is not disabled\n");
            return 1;
        }
/** Threading functions **/

        printf("\nDGEMM & Intel(R) MKL threading\n");

        aux_call_dgemm(n,a,b,c,1); /** DGEMM on N threads (default) **/
        (void) MKL_Domain_Set_Num_Threads(1,MKL_DOMAIN_BLAS);
        aux_call_dgemm(n,a,b,c,1); /** DGEMM on 1 thread **/

/** DGEMM on XXX threads. MKL_DYNAMIC **/

        printf("\nMKL_DYNAMIC experiment\n");

        printf("    Force MKL_DYNAMIC=TRUE\n");
        MKL_Set_Dynamic(1);
        printf("        Set Intel(R) MKL BLAS-N-Threads to 64\n");
        (void) MKL_Domain_Set_Num_Threads(64,MKL_DOMAIN_BLAS);
        aux_print_n_threads();
        aux_call_dgemm(n,a,b,c,1);

        printf("\n    Switch off MKL_DYNAMIC facility\n");
        MKL_Set_Dynamic(0);
        printf("        Set Intel(R) MKL BLAS-N-Threads to 64\n");
        (void) MKL_Domain_Set_Num_Threads(64,MKL_DOMAIN_BLAS);
        aux_print_n_threads();
        aux_call_dgemm(n,a,b,c,1);

        printf("\n    Free DGEMM's arrays\n");
        MKL_free(a);
        MKL_free(b);
        MKL_free(c);
    }  
    seconds_e = dsecnd()-seconds_s;
    printf("\nIntel(R) MKL service functions example finished at%8.4f seconds\n",seconds_e);

    return 0;
}

static void aux_call_dgemm(MKL_INT N, double* A, double* B, double* C, int PRINT_CLOCKS) {

    int      i, j;
    char     transa, transb;
    double   alpha, beta;

    unsigned MKL_INT64  dgemm_s, dgemm_e;
    unsigned MKL_INT64  dgemm_clocks, dgemm_cb, dgemm_ct, dgemm_cu;

    transa = 'N'; transb = 'N';
    alpha = 1.1;  beta = -1.2;

    for ( i=0; i<N; i++ ) {
        for ( j=0; j<N; j++ ) {
            A[j*N+i] = (double)(i+1);
            B[j*N+i] = (double)(-(i+1));
        }
    }

    MKL_Get_Cpu_Clocks(&dgemm_s);
    dgemm(&transa,&transb,&N,&N,&N,&alpha,A,&N,B,&N,&beta,C,&N);
    mkl_get_cpu_clocks(&dgemm_e);

    dgemm_clocks = dgemm_e-dgemm_s;
    if ( PRINT_CLOCKS ) {
        dgemm_cb = dgemm_clocks/1000000;
        dgemm_ct = (dgemm_clocks-(dgemm_cb*1000000))/1000;
        dgemm_cu = (dgemm_clocks-(dgemm_cb*1000000)-(dgemm_ct*1000));
        printf("    DGEMM (%4d) on %2d thread(s):%8.3d.%3d.%3d clocks\n",
               (int)N,MKL_Domain_Get_Max_Threads(MKL_DOMAIN_BLAS),(int)dgemm_cb,(int)dgemm_ct,(int)dgemm_cu);
    }
}

static void aux_print_n_threads(void) {

    if ( MKL_Get_Dynamic() == 0 ) {
        printf("\n        MKL_DYNAMIC : FALSE\n");
    } else {
        printf("\n        MKL_DYNAMIC : TRUE\n");
    }

    printf("        Intel(R) MKL Number of THREADS:  ALL  BLAS FFT PARDISO VML\n");
    printf("                              %4d %4d %4d %4d %4d\n",
           MKL_Get_Max_Threads(),
           MKL_Domain_Get_Max_Threads(MKL_DOMAIN_BLAS),
           MKL_Domain_Get_Max_Threads(MKL_DOMAIN_FFT),
           MKL_Domain_Get_Max_Threads(MKL_DOMAIN_PARDISO),
           MKL_Domain_Get_Max_Threads(MKL_DOMAIN_VML));

    return;
}

File: Examples/intel_examples/examples_core_f/servicefuncsf/source/servicefuncs.f
!===============================================================================
! Copyright 2008-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!      mklservicefunctions example program demonstrate a lot of
!      Intel(R) MKL service functions.
!*******************************************************************************

      program mklservicefunctions

      include "mkl_service.fi"  !! Interfaces for Intel(R) MKL service functions
      include "mkl_lapack.fi"   !! Here is the interface for DSECND

      real*8     freq, SECONDS_S, SECONDS_E

      REAL*8     A,B,C  !! DGEMM's arrays
      POINTER    (A_PTR,A(1)), (B_PTR,B(1)), (C_PTR,C(1))

      INTEGER    N, I

#ifdef _IA32
      integer*4 ALLOC_SIZE, NUM_ELEM, SIZE_ELEM
#else
      integer*8 ALLOC_SIZE, NUM_ELEM, SIZE_ELEM
#endif

      integer*4 NTHRS
      integer*8 ALLOCATED_BYTES
      integer*4 ALLOCATED_BUFFERS
      integer*4 ALIGNMENT

      character*198    buf

      N = 1000
      ALIGNMENT = 128

! Information and defaults

      SECONDS_S = DSECND()

      write(*,'(/A)') 'Intel(R) MKLservice functions example started'
      write(*,'(/A)') 'Intel(R) MKLrelease version:'

      CALL mkl_get_version_string(buf)
      write(*,'(a)') buf

      write(*,'(/A)') 'Information and defaults'

      freq = mkl_get_cpu_frequency()
      write(*,'(A,F8.4,A)') '    Current CPU frequency:', freq,'GHz'

      freq = mkl_get_max_cpu_frequency()
      write(*,'(A,F8.4,A)') '    Maximum CPU frequency:', freq,'GHz'

      freq = mkl_get_clocks_frequency()
      write(*,'(A,F8.4,A)') '    Frequency:', freq,'GHz'

      CALL aux_print_n_threads()

! Memory functions
      ALLOCATED_BYTES = MKL_PEAK_MEM_USAGE(MKL_PEAK_MEM_ENABLE)
      write(*,'(/A)') 'Memory functions'
      write(*,'(A)')  '    Allocate DGEMM''s arrays'
      NUM_ELEM = N*N
      SIZE_ELEM = 8
      ALLOC_SIZE = NUM_ELEM*SIZE_ELEM
      A_PTR = MKL_MALLOC(ALLOC_SIZE,ALIGNMENT)
      if(A_PTR.eq.0) then
         write(*,*) 
     $     'Cannot allocate pointer array of the length ',
     $     ALLOC_SIZE
         stop 1
      endif
      B_PTR = MKL_MALLOC(NUM_ELEM,ALIGNMENT) ! Allocates a memory buffer of smaller size to realloc it to appropriate size later
      if(B_PTR.eq.0) then
         write(*,*)
     $     'Cannot allocate pointer array of the length ',
     $     NUM_ELEM
         stop 1
      endif
      C_PTR = MKL_CALLOC(NUM_ELEM,SIZE_ELEM,ALIGNMENT)
      if(C_PTR.eq.0) then
         write(*,*)
     $     'Cannot allocate pointer array of the length ',
     $     NUM_ELEM
         stop 1
      endif

      B_PTR = MKL_REALLOC(B_PTR,ALLOC_SIZE)
      if(B_PTR.eq.0) then
         write(*,*)
     $     'Cannot reallocate pointer array of the length ',
     $     ALLOC_SIZE
         stop 1
      endif

      write(*,'(A)') '    CALL DGEMM'
      CALL aux_call_dgemm(N,A,B,C,0)
      write(*,'(A)') '    ...Done'

      ALLOCATED_BYTES = MKL_PEAK_MEM_USAGE(MKL_PEAK_MEM)
      write(*,'(A,I8,A)')
     $  '    Peak memory allocated by Intel(R) MKL allocator :',
     $  ALLOCATED_BYTES, ' bytes.'

      ALLOCATED_BYTES = MKL_MEM_STAT(ALLOCATED_BUFFERS)
      write(*,'(A,I8,A,I3,A)')
     $  '    Currently allocated by Intel(R) MKL allocator :',
     $  ALLOCATED_BYTES, ' bytes in',
     $  ALLOCATED_BUFFERS,' buffers.'

      CALL MKL_FREE_BUFFERS()
      ALLOCATED_BYTES = mkl_mem_stat(ALLOCATED_BUFFERS)
      write(*,'(A)') '    After MKL_FREE_BUFFERS was called:'
      write(*,'(A,I8,A,I3,A)')
     $  '    Currenlt allocated by Intel(R) MKL allocator :',
     $  ALLOCATED_BYTES, ' bytes in',
     $  ALLOCATED_BUFFERS,' buffers.'
      if (MKL_PEAK_MEM_USAGE(MKL_PEAK_MEM_DISABLE)<0) then
        write(*,'(/A)') 'Peak memory statistics is not disabled'
      end if

! Threading functions

      write(*,'(/A)') 'DGEMM & Intel(R) MKL threading'

! DGEMM on N thread (default)

      CALL aux_call_dgemm(N,A,B,C,1)

! DGEMM on 1 thread

      NTHRS = 1
      i = MKL_DOMAIN_SET_NUM_THREADS(NTHRS,MKL_DOMAIN_BLAS)
      CALL aux_call_dgemm(N,A,B,C,1)

! DGEMM on XXX threads. MKL_DYNAMIC

      write(*,'(/A)') 'MKL_DYNAMIC experiment'

      write(*,'(A)')  '    Force MKL_DYNAMIC=TRUE'
      CALL MKL_SET_DYNAMIC(MKL_DYNAMIC_TRUE)

      write(*,'(A)')  '        Set Intel(R) MKL BLAS-N-Threads to 64'
      NTHRS = 64
      i = MKL_DOMAIN_SET_NUM_THREADS(NTHRS,MKL_DOMAIN_BLAS)
      CALL aux_print_n_threads()
      CALL aux_call_dgemm(N,A,B,C,1)

      write(*,'(/A)') '    Switch off MKL_DYNAMIC facility'
      CALL MKL_SET_DYNAMIC(MKL_DYNAMIC_FALSE)

      write(*,'(A)')  '        Set Intel(R) MKL BLAS-N-Threads to 64'
      i = MKL_DOMAIN_SET_NUM_THREADS(NTHRS,MKL_DOMAIN_BLAS)
      CALL aux_print_n_threads()
      CALL aux_call_dgemm(N,A,B,C,1)

! Free all work arrays

      write(*,'(/A)') '    Free DGEMM''s arrays'
      CALL MKL_FREE(A_PTR)
      CALL MKL_FREE(B_PTR)
      CALL MKL_FREE(C_PTR)

      SECONDS_E = DSECND()
      SECONDS_E = (SECONDS_E-SECONDS_S)

      write(*,'(/A,F8.4,A)')
     $  'Intel(R) MKLservice functions example finished at',
     $   SECONDS_E,' seconds'

      STOP
      END

      SUBROUTINE aux_call_dgemm(N,A,B,C,PRINT_CLOCKS)

      include 'mkl_service.fi'
      include 'mkl_blas.fi'

      INTEGER    N, PRINT_CLOCKS, I, J
      REAL*8     A(N,N),B(N,N),C(N,N)
      REAL*8     ALPHA, BETA
      integer*8  DGEMM_S, DGEMM_E
      integer*8  DGEMM_CLOCKS, DGEMM_CB, DGEMM_CT, DGEMM_CU

      ALPHA = 1.1; BETA = -1.2
      DO I=1,N
      DO J=1,N
        A(I,J) = I
        B(I,J) = -I
      END DO
      END DO

      CALL MKL_GET_CPU_CLOCKS(DGEMM_S)
      CALL DGEMM('N','N',N,N,N,ALPHA,A,N,B,N,BETA,C,N);
      CALL mkl_get_cpu_clocks(DGEMM_E)
      DGEMM_CLOCKS = DGEMM_E-DGEMM_S

      if (PRINT_CLOCKS == 1) THEN
          DGEMM_CB = DGEMM_CLOCKS/1000000;
          DGEMM_CT = (DGEMM_CLOCKS-(DGEMM_CB*1000000))/1000
          DGEMM_CU = (DGEMM_CLOCKS-(DGEMM_CB*1000000)-(DGEMM_CT*1000))
          write(*,'(A,I4,A,I2,A,I8.3,A,I3.3,A,I3.3,A)')
     $      '    DGEMM (',N,') on ',
     $      MKL_Domain_Get_Max_Threads(MKL_DOMAIN_BLAS), ' thread(s):',
     $      DGEMM_CB,'.',DGEMM_CT,'.',DGEMM_CU,' clocks'
      end if

      RETURN
      END

      SUBROUTINE aux_print_n_threads()

      include 'mkl_service.fi'

      if (MKL_GET_DYNAMIC() == 0) then
          write(*,'(/A,I1)') '        MKL_DYNAMIC : FALSE'
      else
          write(*,'(/A,I1)') '        MKL_DYNAMIC : TRUE'
      end if

      write(*,'(A)')
     $  '   Intel(R) MKL Number of THREADS: ALL BLAS FFT PARDISO VML'
      write(*,'(A,5I5)') '                             ',
     $      MKL_Get_Max_Threads(),
     $      MKL_Domain_Get_Max_Threads(MKL_DOMAIN_BLAS),
     $      MKL_Domain_Get_Max_Threads(MKL_DOMAIN_FFT),
     $      MKL_Domain_Get_Max_Threads(MKL_DOMAIN_PARDISO),
     $      MKL_Domain_Get_Max_Threads(MKL_DOMAIN_VML)

      RETURN
      END
